### PR TITLE
Adopt B95 SoC

### PR DESCRIPTION
**Pull request by liucoldstar**
Adaptation for B95 SoC with correct HAL drivers and libs

### CODE REVIEW

File: drivers/gpio/gpio_b9x.c
Comment (s07641069) [inline drivers/gpio/gpio_b9x.c:1]: Look like needs to be reworked. This a common file for B9X SoCs

File: soc/riscv/riscv-privilege/telink_b9x/soc.c
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c:1]: remove space changes

**Comment by damien0x0023** [comment]
Comment Test (no actual meaning)

File: soc/riscv/riscv-privilege/telink_b9x/soc.c
Comment (liucoldstar) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c:1]: Will do.

File: drivers/gpio/gpio_b9x.c
Comment (liucoldstar) [reply drivers/gpio/gpio_b9x.c:1]: Done.

File: drivers/gpio/gpio_b9x.c [hunk 1]
```diff
@@ -325,11 +369,19 @@ static void gpio_b9x_up_down_res_set(volatile struct gpio_b9x_t *gpio,
 
 	pin = BIT(pin);
 	val = up_down_res & 0x03;
+	#if CONFIG_SOC_RISCV_TELINK_B92
```
Comment (s07641069) [inline drivers/gpio/gpio_b9x.c]: do we need to add B91 SoC here as well?

File: drivers/gpio/gpio_b9x.c [hunk 2]
```diff
@@ -809,6 +873,18 @@ static void gpio_b9x_irq_connect_4(void)
 }
 #endif
 
+/* If instance 4 is present and has interrupt enabled, connect IRQ */
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 5
```
Comment (s07641069) [inline drivers/gpio/gpio_b9x.c:869]: do we need this part of code for all SoCs or only for B95?

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 3]
```diff
@@ -294,7 +358,7 @@ static int soc_b9x_check_flash(void)
 
 	return 0;
 }
+#endif
 
 SYS_INIT(soc_b9x_init, PRE_KERNEL_1, 0);
-
-SYS_INIT(soc_b9x_check_flash, POST_KERNEL, 0);
+/* SYS_INIT(soc_b9x_check_flash, POST_KERNEL, 0); */
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: same here, disable only for b95 soc

**Comment by s07641069** [comment]
Create PR for HAL if there are changes.
And in case if specific HAL needed - change hal revision in west.yml into pull link instead of hash for PR verification

**Comment by andriy-bilynskyy** [comment]
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_telink | https://github.com/telink-semi/hal_telink/commit/0f03ecc893bd5818338ebccf860474310099905e (develop) | https://github.com/telink-semi/hal_telink/pull/77 | [telink-semi/hal_telink#77/files](https://github.com/telink-semi/hal_telink/pull/77/files) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

**Comment by s07641069** [comment]
Still minor changes required described above.
Try to fix all CI jobs for B91 and B92, this will make SoC dependency more clear.

**Comment by haiwentelink** [comment]
For the driver update ,the clk rst and clk en will do in the sys.c , so don't need to do the rst and en in the drvier .

File: .github/workflows/telink-b91-build.yaml
Comment (s07641069) [inline .github/workflows/telink-b91-build.yaml:1]: what is the reason to modify B91 yaml file?

File: .github/workflows/telink-b92-build.yaml
Comment (s07641069) [inline .github/workflows/telink-b92-build.yaml:1]: Same here
what is the reason to modify B92 yaml file?

File: .github/workflows/telink-b95-build.yaml
Comment (s07641069) [inline .github/workflows/telink-b95-build.yaml:1]: in case if B91 and B92 yaml files were modified by mistake - just revert B91 & B 92 yaml file and will be good to make B95 as mach as similar to B91/B92 yaml files

File: drivers/gpio/gpio_b9x.c [hunk 4]
```diff
@@ -440,9 +490,9 @@ static int gpio_b9x_pin_configure(const struct device *dev,
 
 	/* Set GPIO init state if defined to avoid glitches */
 	if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
-		gpio->output |= BIT(pin);
+		GPIO_SET_HIGH_LEVEL(gpio, pin);
 	} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
-		gpio->output &= ~BIT(pin);
+		GPIO_SET_LOW_LEVEL(gpio, pin);
 	}
```
Comment (s07641069) [inline drivers/gpio/gpio_b9x.c:496]: double check this change, not sure if they are correct

File: drivers/hwinfo/hwinfo_b9x.c [hunk 5]
```diff
@@ -13,9 +13,13 @@ ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
 	uint32_t flash_mid = 0;
 	uint8_t uid[16];
 
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 	flash_mid = flash_read_mid();
-
 	flash_read_mid_uid_with_check((unsigned int *)&flash_mid, uid);
+#else
```
Comment (s07641069) [inline drivers/hwinfo/hwinfo_b9x.c]: use #elif CONFIG_SOC_RISCV_TELINK_B95

File: drivers/ieee802154/ieee802154_b9x.c [hunk 6]
```diff
@@ -1042,8 +1045,6 @@ static int b9x_stop(const struct device *dev)
 		}
 		riscv_plic_irq_disable(DT_INST_IRQN(0) - CONFIG_2ND_LVL_ISR_TBL_OFFSET);
 		rf_set_tx_rx_off();
-		rf_baseband_reset();
-		rf_reset_dma();
```
Comment (s07641069) [inline drivers/ieee802154/ieee802154_b9x.c]: what is the reason to remove?
As I remember this code required for B91 & B92

File: drivers/ieee802154/ieee802154_b9x_frame.c [hunk 7]
```diff
@@ -612,7 +618,12 @@ ieee802154_b9x_crypto_ecb(
 	const uint8_t inp[IEEE802154_CRYPTO_LENGTH_AES_BLOCK],
 	uint8_t out[IEEE802154_CRYPTO_LENGTH_AES_BLOCK])
 {
+#ifdef CONFIG_SOC_RISCV_TELINK_B95
+	(void)ske_lp_crypto(SKE_ALG_AES_128, SKE_MODE_ECB,
+	SKE_CRYPTO_ENCRYPT, (uint8_t *)key, 0, NULL, (uint8_t *)inp, out, 16);
+#else
```
Comment (s07641069) [inline drivers/ieee802154/ieee802154_b9x_frame.c]: lets use everywhere one style as above:
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
.....
#elif CONFIG_SOC_RISCV_TELINK_B95
....
#endif

File: drivers/serial/uart_b9x.c [hunk 8]
```diff
@@ -187,9 +187,13 @@ static void uart_b9x_cal_div_and_bwpc(uint32_t baudrate, uint32_t pclk,
 static void uart_b9x_init(volatile struct uart_b9x_t *uart, uint16_t divider,
 			  uint8_t bwpc, uint8_t parity, uint8_t stop_bit)
 {
+#if CONFIG_SOC_RISCV_TELINK_B95
+	uart->ctrl0 = ((uart->ctrl0 & (~FLD_UART_BPWC_O)) | bwpc);
+#else
+	uart->ctrl0 = bwpc;
+#endif
```
Comment (s07641069) [inline drivers/serial/uart_b9x.c]: same here lets use one style for #if SoCs as above
and what is the reason to move this upper line "divider = ..." ?

File: samples/net/openthread/cli/boards/tlsr9258a.overlay [hunk 9]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2023 Telink Semiconductor
```
Comment (s07641069) [inline samples/net/openthread/cli/boards/tlsr9258a.overlay]: 2024

File: samples/net/openthread/coprocessor/boards/tlsr9258a.conf [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# Telink B92 TLSR9258A
```
Comment (s07641069) [inline samples/net/openthread/coprocessor/boards/tlsr9258a.conf]: B95

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 11]
```diff
@@ -8,7 +8,9 @@
 #include <clock.h>
 #include <gpio.h>
 #include <ext_driver/ext_pm.h>
+/*
 #include "rf.h"
+*/
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: why this include is disabled?

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 12]
```diff
@@ -280,14 +354,16 @@ static int soc_b9x_check_flash(void)
...
+#endif
 	default:
 		break;
 	}
 
 	if (hw_flash_size < dts_flash_size) {
+		/**
 		printk("!!! flash error: expected (.dts) %u, actually %u\n",
 			dts_flash_size, hw_flash_size);
+		**/
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: why this error print is disabled?

File: drivers/gpio/gpio_b9x.c (hunk 2 above)
Comment (s07641069) [reply drivers/gpio/gpio_b9x.c:869]: if there no impact into B91&B92 we can leave as it is, just double check

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 11 above)
Comment (s07641069) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: use #if for B95 only if this is temporary needed

**Review by s07641069 — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_b9x.c (hunk 1 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c]: For B91 don't have PROT(F),so will not add

File: drivers/gpio/gpio_b9x.c (hunk 2 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c:869]: will add B95 define for the new add part .

File: .github/workflows/telink-b91-build.yaml
Comment (haiwentelink) [reply .github/workflows/telink-b91-build.yaml:1]: will revert B91

File: .github/workflows/telink-b92-build.yaml
Comment (haiwentelink) [reply .github/workflows/telink-b92-build.yaml:1]: will revert B92

File: .github/workflows/telink-b95-build.yaml
Comment (haiwentelink) [reply .github/workflows/telink-b95-build.yaml:1]: change back the B91,B92 , but the B95 not support so many samples , we can only keep the sample we already tested , and will add the other samples after we add new code but we use the same style as the B92

File: drivers/gpio/gpio_b9x.c (hunk 4 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c:496]: already changed

File: drivers/serial/uart_b9x.c (hunk 8 above)
Comment (haiwentelink) [reply drivers/serial/uart_b9x.c]: actually it's meaning less , but i think if need ,we shoud move the code like that(related code works together)
/* config clock */
	divider = divider | FLD_UART_CLK_DIV_EN;
        uart->clk_div = divider;

File: drivers/gpio/gpio_b9x.c (hunk 1 above)
Comment (s07641069) [reply drivers/gpio/gpio_b9x.c]: checking if port available for soc is done by macros IS_PORT_F
Here is my proposal to clean up this part of code:
```
	#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	if (IS_PORT_F(gpio)) {
		analog_reg = 0x23 + ((pin & 0xf0) ? 1 : 0);
	} else {
		analog_reg = 0x0e + (GET_PORT_NUM(gpio) << 1) + ((pin & 0xf0) ? 1 : 0);
	}
	#elif CONFIG_SOC_RISCV_TELINK_B95
	if ((IS_PORT_F(gpio)) || (IS_PORT_G(gpio))) {
		return;
	} else {
		analog_reg = 0x17 + (GET_PORT_NUM(gpio) << 1) + ((pin & 0xf0) ? 1 : 0);
	}
	#endif
```

File: drivers/ieee802154/ieee802154_b9x.c (hunk 6 above)
Comment (haiwentelink) [reply drivers/ieee802154/ieee802154_b9x.c]: For the initial version not have these interface ,have already add .

File: drivers/ieee802154/ieee802154_b9x_frame.c [hunk 13]
```diff
@@ -603,7 +603,15 @@ b9x_ieee802154_get_data(const uint8_t *payload,
  * Cryptography functionality
  */
 
+#ifdef CONFIG_IEEE802154_2015
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
 #include <aes.h>
+#elif CONFIG_SOC_RISCV_TELINK_B95
+#include <ske.h>
+#include <ske_portable.h>
+#else
```
Comment (s07641069) [inline drivers/ieee802154/ieee802154_b9x_frame.c]: we haven't other SoCs in B9X family, why we need here #else  ?
looks like can be removed

**Review by s07641069 — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_b9x.c (hunk 1 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c]: yes , I have checked it will be OK , and the code will be clean .

File: drivers/ieee802154/ieee802154_b9x_frame.c (hunk 7 above)
Comment (s07641069) [reply drivers/ieee802154/ieee802154_b9x_frame.c]: added another comment, looks like #else needs to be removed now

File: drivers/ieee802154/ieee802154_b9x_frame.c (hunk 13 above)
Comment (haiwentelink) [reply drivers/ieee802154/ieee802154_b9x_frame.c]: OK, i will check with zhenghuan , seems can be removed.

File: drivers/gpio/gpio_b9x.c (hunk 4 above)
Comment (s07641069) [reply drivers/gpio/gpio_b9x.c:496]: somehow I don't see any change in latest commit
have you made some change or you wants to leave as it is?

File: drivers/gpio/gpio_b9x.c (hunk 2 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c:871]: I look into the code ,and find it's changed by Dmytro , for B95 , the io set high and low will use different register .

#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output, pin, 0)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#elif CONFIG_SOC_RISCV_TELINK_B95
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output_clr, pin, 1)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#endif
 and the code GPIO_SET_LOW_LEVEL(gpio, pin); is equal to the WRITE_BIT(gpio->output, pin, 0);
also the WRITE_BIT(gpio->output, pin, 0); is equal to gpio->output &= ~BIT(pin);

#define WRITE_BIT(var, bit, set) \
	((var) = (set) ? ((var) | BIT(bit)) : ((var) & ~BIT(bit)))
so if we want to write gpio , we should use the GPIO_SET_LOW_LEVEL or GPIO_SET_HIGH_LEVEL, can not directly operate WRITE_BIT or directly use register .

File: drivers/gpio/gpio_b9x.c (hunk 4 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c:496]: I look into the code ,and find it's changed by Dmytro , for B95 , the io set high and low will use different register .

```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output, pin, 0)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#elif CONFIG_SOC_RISCV_TELINK_B95
#define GPIO_SET_LOW_LEVEL(gpio, pin)   WRITE_BIT(gpio->output_clr, pin, 1)
#define GPIO_SET_HIGH_LEVEL(gpio, pin)  WRITE_BIT(gpio->output, pin, 1)
#endif
 and the code GPIO_SET_LOW_LEVEL(gpio, pin); is equal to the WRITE_BIT(gpio->output, pin, 0);
also the WRITE_BIT(gpio->output, pin, 0); is equal to gpio->output &= ~BIT(pin);

#define WRITE_BIT(var, bit, set) \
	((var) = (set) ? ((var) | BIT(bit)) : ((var) & ~BIT(bit)))
```
      so if we want to write gpio , we should use the GPIO_SET_LOW_LEVEL or GPIO_SET_HIGH_LEVEL, can not directly operate WRITE_BIT or directly use register .

File: drivers/ieee802154/ieee802154_b9x_frame.c (hunk 13 above)
Comment (haiwentelink) [reply drivers/ieee802154/ieee802154_b9x_frame.c]: already update code

File: soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc [hunk 14]
```diff
@@ -38,7 +38,7 @@ config TELINK_B9x_HWDSP
 config TELINK_B9X_MBEDTLS_HW_ACCELERATION
 	bool "Use Telink B9x platform mbedtls HW acceleration"
 	depends on SOC_SERIES_RISCV_TELINK_B9X && MBEDTLS
-	default y
+	default n
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc]: This will disable HW crypto for all B9X SoCs
Revert this change and if you need to disable only for B95, jut set "CONFIG_TELINK_B9X_MBEDTLS_HW_ACCELERATION=n" in some conf or Kconfig file of B95 SoC (board)

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 15]
```diff
@@ -20,7 +20,11 @@
 #endif
 
 /* Software reset defines */
+#ifdef CONFIG_SOC_RISCV_TELINK_B95
+#define reg_reset                   REG_ADDR8(0x14082f)
+#else
 #define reg_reset                   REG_ADDR8(0x1401ef)
+#endif
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: Again use same style:
```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	...
#elif CONFIG_SOC_RISCV_TELINK_B95
	...
#endif
```

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 16]
```diff
@@ -58,14 +64,14 @@
...
-		#define POWER_MODE      DCDC_1P4_LDO_2P0
+		#define POWER_MODE      DCDC_0P94_LDO_1P8
 	#elif DT_ENUM_IDX(DT_NODELABEL(power), power_mode) == 2
-		#define POWER_MODE      DCDC_1P4_DCDC_2P0
+		#define POWER_MODE      DCDC_0P94_DCDC_1P8
 	#else
-		#error "Wrong value for power-mode parameter"
-	#endif
+	#error "Wrong value for power-mode parameter"
+#endif
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: check tabs, for this #endif it is missed

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 17]
```diff
@@ -226,32 +261,69 @@ void soc_b9x_restore(void)
...
+	case CLK_120MHZ:
+		PLL_240M_CCLK_120M_HCLK_60M_PCLK_60M_MSPI_48M;
+		break;
+#endif
 	}
 }
 
+static inline uint32_t read_flash_mid(void)
+{
+#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
```
Comment (s07641069) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c:320]: again one style
```
#if CONFIG_SOC_RISCV_TELINK_B91 || CONFIG_SOC_RISCV_TELINK_B92
	...
#elif CONFIG_SOC_RISCV_TELINK_B95
	...
#endif
```

**Review by s07641069 — CHANGES_REQUESTED** [review]

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 15 above)
Comment (haiwentelink) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: OK, will check all the define of the CONFIG_SOC_RISCV_TELINK_B95, to make sure all the code to follow the same style .

File: soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc (hunk 14 above)
Comment (s07641069) [reply soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc]: TELINK_B9X_MBEDTLS_HW_ACCELERATION=y required for Matter
there two cases:
1. Disable only for B95 and will need to rework Matter. Anyway will need in near future.
2. Revert to Y and create jira ticket for B95 to Adopt/fix and test HW crypto for B95. Better to do this.

File: drivers/gpio/gpio_b9x.c (hunk 4 above)
Comment (haiwentelink) [reply drivers/gpio/gpio_b9x.c:496]: already test sample/basic/blinky and sample/basic/button on B91 ,and also measured the voltage, i think there will be no problem .

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 17 above)
Comment (haiwentelink) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c:320]: fixed

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 16 above)
Comment (haiwentelink) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: fixed

File: soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc (hunk 14 above)
Comment (haiwentelink) [reply soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc]: already revert in Kconfig.soc , and set the TELINK_B9X_MBEDTLS_HW_ACCELERATION=y in the tlsr9258a_defconfig

Comment (pulin1103) [reply soc/riscv/riscv-privilege/telink_b9x/Kconfig.soc]: About CONFIG_TELINK_B9X_MBEDTLS_HW_ACCELERATION, I have create a ticket ZEPHYR-503, see my comment in it.

**Review by s07641069 — APPROVED** [review]
